src/braket/aws/aws_device.py

```diff
@@ -622,3 +679,265 @@ def get_device_region(device_arn: str) -> str:
+            GateCalibrations: the calibration data for the device.
+        """
+        if self.gate_calibrations_href is not None:
+            try:
+                with urllib.request.urlopen(self.gate_calibrations_href.split("?")[0]) as f:
+                    json_calibration_data, fidelities = self._parse_calibration_json(
+                        json.loads(f.read().decode("utf-8"))
+                    )
+                    self._gate_calibrations_timestamp = datetime.now()
+                    return GateCalibrations(json_calibration_data, fidelities)
+            except urllib.error.URLError:
+                raise urllib.error.URLError(f"Unable to reach {self.gate_calibrations_href}")
+        else:
+            return None
+
+    def _parse_constant_waveform(self, w: Dict) -> Tuple[float, complex]:
+        length = iq = None
+        for val in w["arguments"]:
+            if val["name"] == "length":
+                length = (
+                    float(val["value"])
+                    if is_float(val["value"])
+                    else FreeParameterExpression(val["value"])
+                )
+            if val["name"] == "iq":
+                iq = (
+                    complex(val["value"])
+                    if is_float(val["value"])
+                    else FreeParameterExpression(val["value"])
+                )
+        return length, iq
+
+    def _parse_waveforms(self, waveforms_json: str) -> Dict:
+        waveforms = dict()
+        for waveform in waveforms_json:
+            w = waveforms_json[waveform]
+            wave_id = w["waveformId"]
+            waveform_parameters = {"id": wave_id}
+            if "amplitudes" in w.keys():
+                complex_amplitudes = [complex(i[0], i[1]) for i in w["amplitudes"]]
+                waveforms[wave_id] = ArbitraryWaveform(complex_amplitudes, wave_id)
+            elif w["name"] == "drag_gaussian":
+                for val in w["arguments"]:
+                    waveform_parameters[val["name"]] = (
+                        float(val["value"])
+                        if is_float(val["value"])
+                        else FreeParameterExpression(val["value"])
+                    )
+                waveforms[wave_id] = DragGaussianWaveform(**waveform_parameters)
+            elif w["name"] == "gaussian":
+                for val in w["arguments"]:
+                    waveform_parameters[val["name"]] = (
+                        float(val["value"])
+                        if is_float(val["value"])
+                        else FreeParameterExpression(val["value"])
+                    )
+                waveforms[wave_id] = GaussianWaveform(**waveform_parameters)
+            elif w["name"] == "constant":
+                length = iq = self._parse_constant_waveform(w)
+                waveforms[wave_id] = ConstantWaveform(length, iq)
+            else:
+                raise ValueError(f"The waveform {wave_id} of cannot be constructed")
+        return waveforms
```

Comment:
None of this parsing logic belongs in this class.

Reply:
Cleaned up this and instruction parsing logic.